File: C#/Task450.cs
using System;
using System.Collections.Generic;

public class Task450
{
    public static void Main(string[] args)
    {
        // список навчальних предметів
        List<string> subjects = new List<string>
        {
            "Алгеба", "Хімія", "Фізика", "Інформатика"
        };
        // список студентів
        List<string> students = new List<string>
        {
            "Вовк", "Заєць", "Ведмідь", "Лис", "Лось"
        };
        // результати сесії: студент -> (предмет -> оцінка)
        Dictionary<string, Dictionary<string, int>> session =
            new Dictionary<string, Dictionary<string, int>>();

        Random rnd = new Random();
        foreach (var student in students)
        {
            var marks = new Dictionary<string, int>();

            foreach (var subject in subjects)
            {
                marks[subject] = 50 + rnd.Next(51);
            }
            session[student] = marks;
        }
        // друк результатів
        Console.WriteLine("Успішність студентів:");
        foreach (var student in session.Keys)
        {
            Console.Write(student + " : ");
            foreach (var subject in session[student].Keys)
            {
                Console.Write(subject + " - " +
                              session[student][subject] + "; ");
            }
            Console.WriteLine();
        }
        // формуємо рейтинг
        Console.WriteLine("Невпорядкований рейтинг");
        Dictionary<string, double> rating =
            new Dictionary<string, double>();
        foreach (var student in session.Keys)
        {
            int sum = 0;
            int count = session[student].Count;
            foreach (var subject in session[student].Keys)
            {
                sum += session[student][subject];
            }
            double rate = (double)sum / count;
            rating[student] = rate;
            Console.WriteLine(student + " : " + rate);
        }
        // формуємо список студентів за рейтингом
        List<string> ratingList = new List<string>();
        foreach (var student in rating.Keys)
        {
            int id = 0;
            foreach (var surname in ratingList)
            {
                if (rating[surname] < rating[student])
                {
                    break;
                }
                id++;
            }
            ratingList.Insert(id, student);
        }
        // друк рейтингу
        Console.WriteLine("Рейтинг студентів ");
        foreach (var student in ratingList)
        {
            Console.WriteLine(student + " : " + rating[student]);
        }
    }
}

File: Java/Task450.java
import java.util.*;

public class Task450 {
    public static void main(String[] args) {
        // список навчальних предметів
        List<String> subjects = Arrays.asList("Алгеба",
                     "Хімія", "Фізика", "Інформатика");
        // список студентів
        List<String> students = new ArrayList<>(Arrays
       .asList("Вовк", "Заєць", "Ведмідь", "Лис", "Лось"));
        // результати сесії: студент -> (предмет -> оцінка)
        Map<String, Map<String, Integer>> session = 
                                  new LinkedHashMap<>();
        Random rnd = new Random();
        for (String student : students) {
            Map<String, Integer> marks =
                              new LinkedHashMap<>();
            for (String subject : subjects) {
                marks.put(subject, 50 + rnd.nextInt(51));
            }
            session.put(student, marks);
        }
        // друк результатів
        System.out.println("Успішність студентів:");
        for (String student : session.keySet()) {
            System.out.print(student + " : ");
            for (String subject :
                   session.get(student).keySet()) {
                System.out.print(subject + " - " +
              session.get(student).get(subject) + "; ");
            }
            System.out.println();
        }
        // формуємо рейтинг
        System.out.println("Невпорядкований рейтинг");
        Map<String, Double> rating = new LinkedHashMap<>();
        for (String student : session.keySet()) {
            int sum = 0;
            int count = session.get(student).size();
            for (String subject : 
                 session.get(student).keySet()) {
                sum += session.get(student).get(subject);
            }
            double rate = (double) sum / count;
            rating.put(student, rate);
            System.out.println(student + " : " + rate);
        }
        // формуємо список студентів за рейтингом
        List<String> ratingList = new ArrayList<>();
        for (String student : rating.keySet()) {
            int id = 0;
            for (String surname : ratingList) {
                if (rating.get(surname) <
                            rating.get(student)) {
                    break;
                }
                id++;
            }
            ratingList.add(id, student);
        }
        // друк рейтингу
        System.out.println("Рейтинг студентів ");
        for (String student : ratingList) {
            System.out.println(student + " : " +
                  rating.get(student));
        }
    }
}
